lua/ttt2/libraries/spawn.lua

```diff
@@ -0,0 +1,290 @@
+-- @param [opt]table|Entity filder A table of entities or an entity that should be ignored by this check
+-- @return bollean Teturns if the spawn point is safe
+-- @realm server
+function spawn.IsSpawnPointSafe(pos, force, filter)
+	if not util.IsInWorld(pos) then
+		return false
+	end
+
+	filter = istable(filter) and filter or {filter}
+
+	local posCenter = pos + Vector(0, 0, 0.5 * sizePlayer.z)
+
+	-- Make sure there is enough space around the player
+	local traceWalls = util.TraceHull({
+		start = posCenter,
+		endpos = posCenter,
+		mins = -0.5 * sizePlayer,
+		maxs = 0.5 * sizePlayer,
+		filter = function(ent)
+			if not IsValid(ent) or ent:IsPlayer() then
+				return false
+			end
+
+			if ent:HasPassableCollisionGrup() then
+				return false
+			end
+
+			return true
+		end,
+		mask = MASK_SOLID
+	})
+
+	if traceWalls.Hit then
+		return false
+	end
+
+	-- make sure the new position is on the ground
+	local traceGround = util.TraceLine({
+		start = posCenter,
+		endpos = posCenter - Vector(0, 0, sizePlayer.z),
+		filter = player.GetAll(),
+		mask = MASK_SOLID
+	})
+
+	if not traceGround.Hit then
+		return false
+	end
+
+	local blockingEntities = ents.FindInBox(
+		pos + Vector(-0.5 * sizePlayer.x, -0.5 * sizePlayer.y, 0),
+		pos + sizePlayer
+	)
+
+	for i = 1, #blockingEntities do
+		local blockingEntity = blockingEntities[i]
+
+		if not IsValid(blockingEntity) or not blockingEntity:IsPlayer()
+			or not blockingEntity:IsTerror() or not blockingEntity:Alive()
+		then continue end
+
+		if table.HasValue(filter, blockingEntity) then continue end
+
+		if force then
+			blockingEntity:Kill()
+		else
+			return false
+		end
+	end
+
+	return true
+end
+
+---
+-- Returns a selection of points around a given point.
+-- @param Vector pos The given position
+-- @return table A table of position vectors
+-- @realm server
+function spawn.GetSpawnPointsAroundSpawn(pos)
+	if not pos then return {} end
+
+	local positions = {}
+
+	for i = 1, #spawnPointVariations do
+		positions[#positions + 1] = pos + spawnPointVariations[i] * Vector(sizePlayer.x, sizePlayer.y, 0) * 1.5
+	end
+
+	return positions
+end
+
+---
+-- Tries to make spawn position valid by scanning surrounding area for
+-- valid positions.
+-- @param Vector unsafePos The unsafe spawn position
+-- @return Vector|nil Returns the safe spawn position, nil if none was found
+-- @realm server
+function spawn.MakeSpawnPointSafe(unsafePos)
+	local spawnPoints = spawn.GetSpawnPointsAroundSpawn(unsafePos)
+
+	for i = 1, #spawnPoints do
+		local spawnPoint = spawnPoints[i]
+
+		if not spawn.IsSpawnPointSafe(spawnPoint, false) then continue end
+
+		return spawnPoint
+	end
+end
+
+---
+-- Returns a list of all spawn entities
+-- @param boolean shouldShuffle whether the table should be shuffled
+-- @param boolean forceAll used unless absolutely necessary (includes info_player_start spawns)
```

Comment:
i can't comprehend what you mean by this

Reply:
This is copied from vanilla TTT, but you're right, I'll change it